### ISSUE TITLE
Bug: "Provider produced inconsistent final plan" when executing terraform_data resource triggered immediately after `maas_vm_host_machine` resource

### ISSUE THREAD

**Issue by vishvikkrishnan** [canonical/terraform-provider-maas#224]
I'm using local-execs to work around the provider's limitation of only assigning multiple machines to a single tag (like the `maas_tag resource` does today), and not the other way round. I also use the `triggers_replace` block of the `terraform_data` resource to ensure the tags are force-replaced in case the `maas_vm_host_machine` is replaced during `terraform apply`.

When trying to specify the network_interfaces block for the VM, however, terraform errors out immediately after creating the VMs with the following error:

```shell
maas_vm_host_machine.compose_vms["juju-vm-1"]: Creation complete after 4m4s [id=n7byay]                                                                                                          
╷                                                                                                                                                                                                
│ Error: Provider produced inconsistent final plan                                                                                                                                               
│                                                                                                                                                                                                
│ When expanding the plan for terraform_data.apply_tags["juju-vm-1"] to                                                                                                                          
│ include new values learned so far during apply, provider                                                                                                                                       
│ "terraform.io/builtin/terraform" produced an invalid new value for                                                                                                                             
│ .triggers_replace[0].network_interfaces[0].vlan: was null, but now                                                                                                                             
│ cty.StringVal("").                                                                                                                                                                             
│                                                                                                                                                                                                
│ This is a bug in the provider, which should be reported in the provider's                                                                                                                      
│ own issue tracker.                                                                                                                                                                             
╵                                                                                                                                                                                                
╷                                                                                                                                                                                                
│ Error: Provider produced inconsistent final plan
│ 
│ When expanding the plan for terraform_data.apply_tags["juju-vm-1"] to
│ include new values learned so far during apply, provider
│ "terraform.io/builtin/terraform" produced an invalid new value for
│ .triggers_replace[0].network_interfaces[0].ip_address: was null, but now
│ cty.StringVal("").
│ 
│ This is a bug in the provider, which should be reported in the provider's
│ own issue tracker.
```

This breaks the flow of my module, without running the subsequent `terraform_data` resource in charge of applying tags to the commissioned VM(s). Rerunning the same plan allows TF to properly apply the tags to the VM(s) without any issues. Reporting this as an issue since the error literally mentions that. It looks like a type-error before and after the `maas_vm_host_machine` is applied.

Here is the main.tf file:

```hcl
resource "maas_vm_host_machine" "compose_vms" {
  for_each = var.vms

  hostname = each.key
  vm_host  = each.value.vm_host
  cores    = each.value.vm.cores
  memory   = each.value.vm.memory
  zone     = each.value.vm.zone
  pool     = each.value.vm.pool

  dynamic "storage_disks" {
    for_each = coalesce(each.value.vm.disks, [])
    content {
      size_gigabytes = storage_disks.value.size
      pool           = storage_disks.value.storage_pool
    }
  }
  dynamic "network_interfaces" {
    for_each = coalesce(each.value.vm.interfaces, [])
    content {
      name = network_interfaces.value.name
      fabric = network_interfaces.value.fabric
      subnet_cidr = network_interfaces.value.subnet_cidr
    }
  }
}

resource "terraform_data" "apply_tags" {
  for_each = var.vms

  triggers_replace = [
    maas_vm_host_machine.compose_vms[each.key],
    each.value
  ]

  provisioner "local-exec" {
    <some super-crazy maas cli commands with TF looping string directives so the same commands are run for different tags and different VMs>
  }
}

It should be possible for terraform to run the `terraform_data` resource following `maas_vm_host_machine` without any errors on the first run itself.

**Comment by skatsaounis** [comment]
Hi @vishvikkrishnan ,

tl;dr May I suggest the following change with the hope that it can work with your module's logic? Based on you description I assume that if `id` aka MAAS `system_id` is changed, this is enough for your logic to run. Unless you want to monitor other fields too. If this is the case, please add more fields in the `triggers_replace`.

```diff
triggers_replace = [
-    maas_vm_host_machine.compose_vms[each.key],
+    maas_vm_host_machine.compose_vms[each.key].id,
    each.value
  ]
```
As for the issue itself: The MAAS provider is implemented in Terraform Plugin SDKv2. What you observe is derived from this and more info can be found here: https://developer.hashicorp.com/terraform/plugin/framework/migrating/benefits#fully-exposed-value-states. If you try to plan before apply you can see something like the following:

```bash
# terraform_data.apply_tags["test"] will be created
  + resource "terraform_data" "apply_tags" {
      + id               = (known after apply)
      + triggers_replace = [
          + {
              ...
              + network_interfaces = [
                  + {
                      + fabric      = "fabric-1"
                      + ip_address  = null
                      + name        = "enp5s0"
                      + subnet_cidr = "10.20.0.0/24"
                      + vlan        = null
                    },
                ]
              ...
            },
          + {
              ...
              + interfaces = [
                  + {
                      + fabric      = "fabric-1"
                      + name        = "enp5s0"
                      + subnet_cidr = "10.20.0.0/24"
                    },
                ]
             ...
            },
        ]
    }
```

And if you observe the Terraform state right after apply you can see that the resource is created with `""` on both `vlan` and `ip_address`:

```bash
"resources": [
    {
      "mode": "managed",
      "type": "maas_vm_host_machine",
      "name": "compose_vms",
      "provider": "provider[\"registry.terraform.io/canonical/maas\"]",
      "instances": [
        {...
            "network_interfaces": [
              {
                "fabric": "fabric-1",
                "ip_address": "",
                "name": "enp5s0",
                "subnet_cidr": "10.20.0.0/24",
                "vlan": ""
              }
            ],
...
```

I will proceed with closing the issue as not planned since the re-implementation of the provider with framework is a major change which is not planned for now or near future.
